### PR TITLE
add xmlReadFile and parse_file with options

### PR DESCRIPTION
**Pull request by tkelman**
This is useful for setting options flags, ref http://www.xmlsoft.org/html/libxml-parser.html#xmlReadFile and http://www.xmlsoft.org/html/libxml-parser.html#xmlParserOption

~~TODO: tests and docs~~ done

### CODE REVIEW

**Comment by timholy** [comment]
LGTM

File: src/document.jl [hunk 1]
```diff
@@ -90,6 +90,13 @@ function parse_file(filename::AbstractString)
     XMLDocument(p)
 end
 
+function parse_file(filename::AbstractString, encoding, options::Integer)
+    p = ccall((:xmlReadFile,libxml2), Xptr, (Cstring, Ptr{Cchar}, Cint),
```
Comment (stevengj) [inline src/document.jl]: second argument should be `Cstring` too, since it is presumably a NUL-terminated string (you aren't passing a length).

Comment (stevengj) [inline src/document.jl]: Oh, I see, you want to be able to pass `C_NULL` for `encoding`.  That seems pretty un-Julian to expose.  Maybe just have a separate `parse_file(filename, options)` method?

File: test/parse.jl [hunk 2]
```diff
@@ -21,7 +21,9 @@ docstr = """
 </bookstore>
 """
 
-xdoc = parse_string(docstr)
+for xdoc = (parse_string(docstr),
+            parse_file(joinpath(dirname(@__FILE__), "ex1.xml")),
+            parse_file(joinpath(dirname(@__FILE__), "ex1.xml"), C_NULL, 64)) # 64 == XML_PARSE_NOWARNING
```
Comment (stevengj) [inline test/parse.jl]: add a test case for passing the encoding?

Comment (tkelman) [inline test/parse.jl]: don't actually have a use case for that, would rather this just be a pretty thin wrapper around the C API

File: src/document.jl (hunk 1 above)
Comment (tkelman) [inline src/document.jl]: If I wanted this to be much prettier I'd probably do a keyword-argument version with all the option enums spelled out. I could maybe have called this `read_file` to make it more clearly a direct mapping of the C API, but doesn't seem worth the trouble to have a different argument order.

Comment (stevengj) [inline src/document.jl]: The problem with not using `Cstring` is that you get silent truncation of `filename` strings with embedded NUL chars.

Comment (tkelman) [inline src/document.jl]: Just `encoding` strings with embedded nul's. I'd rather have this API be closer to the extensively documented upstream C API, even if that means a few methods are a little unsafe. There are bigger issues to worry about here, like memory management being kind of a mess without reference counts.

**Pull request merged**